### ISSUE TITLE
CDMs are platform dependant

### ISSUE THREAD

**Issue by GravisZro** [w3c/encrypted-media#157]
A major problem with CDMs is that they are (currently) platform and architecture dependant.  This means that content that would otherwise work on anything will be restricted to only being viewable on the most popular platforms and architectures.  So even if by some miracle a CDM is ported to Linux, it's inconceivable that it would be ported to the Linux PowerPC platform.  Divvying up the internet into supported and unsupported platforms is a step backwards.

**Comment by steelejoe** [comment]
**This is not a technical issue with the specification.**  If your concern is whether a CDM is implementable on any given platform, then answer is that it may depend on the specifics of the Key System being implemented. However for any platform or architecture, it is possible to implement a CDM. The ClearKey key system is the primary example.

**Comment by GravisZro** [comment]
**It is a technical issue with the specification.**  By leaving the platform of the CDM undefined you enable ambiguity which results in incompatibility.

It's just as large an issue as not defining endianness and using multibyte characters.  Sure, webpages working on x86 are great but when they don't work on PowerPC, "whoop"?

CDMs should be platform and architecture agnostic and saying it's not a technical issue with the specification is a lie.

**Comment by mwatson2** [comment]
Strictly, instead of:
> This means that content that would otherwise work on anything will be restricted to only being viewable on the most popular platforms and architectures.
you should say "this means that content that would previously work on any platform supporting a content-specific proprietary plugin (e.g. Flash or Silverlight) will now work only on platforms supporting a proprietary CDM supported by the content provider."

This is true. Whether with respect to interoperability specifically it is a step forward, backward or neutral is debatable. The CDMs have drastically reduced footprint compared to the plugins they replace, so one might speculate that cross-platform support would be less of an overhead for their developers.

A truly cross-platform solution would require standardization of some kind of trusted virtual execution environment, which is way beyond the scope of our work here.

**Comment by GravisZro** [comment]
@mwatson2 
taking a page from steelejoe's playbook, other systems are beyond the scope of this spec.

**Comment by mwatson2** [comment]
@GravisZro indeed. But I'm not understanding your point in noting that.

**Comment by GravisZro** [comment]
@mwatson2 by being outside the scope of the spec it is irrelevant to this conversation and therefore be dismissed which means that you have not made a valid counter-argument.

**Comment by mwatson2** [comment]
@GravisZro I disagree. Many things are outside the scope of the specification by nevertheless relevant to discussion of the specification. In particular, if a specification is designed to replace a pre-existing system then the relative merits of the specification compared to that system are especially relevant.

**Comment by GravisZro** [comment]
@mwatson2 Well if we are going to talk about the reality, there is nothing wrong with simply not using EME.  It is actively harmful to the common good as it actively excludes individuals who do not in a supported platform, architecture or country from being able to view content.  This is in direct conflict with the [W3C Mission's Design Principles](https://www.w3.org/Consortium/mission).

> The social value of the Web is that it enables human communication, commerce, and opportunities to share knowledge. One of W3C's primary goals is to make these benefits available to all people, **whatever their hardware, software**, network infrastructure, native language, culture, **geographical location**, or physical or mental ability.

CDMs violate these design principles.  To fix this problem, CDMs must be hardware and platform agnostic.  I would also argue that CDMs restricting access based on geographic location is a problem.

**EME is no better than the terrible things corporations have made** because it has all the same hardware/software/geographic restrictions.  The only way to make it _any_ better is to remove the hardware/software restrictions.

**Comment by mwatson2** [comment]
There is no way, as far as I know, for EME CDMs to impose geographic restrictions (they do not have network access or other access to geographic information). EME CDMs also have very different security and privacy properties compared to the previous solutions (I assume this is what you mean by the "terrible things corporations have made").

So, I disagree with your last sentence.

As for addressing the issue of platform-dependence, "not using EME" would leave people only with those previous solutions and so I do not see that as a viable answer.

I'm open to alternatives that improve the situation but in the absence of any viable suggestions, I think we should close this one.

**Comment by GravisZro** [comment]
it's simple, make CDMs run in a virtual environment.

**Comment by mwatson2** [comment]
> sure there is. upon requesting a license, check if the IP is from a country you support, if not, no reply.

That would be the license server imposing the restriction, not the CDM. You can do that with or without EME.

> i do have a suggestion: make CDMs run in a virtual environment. similar to java, the environment can then be ported then all CDMs can run on any platform or architecture.

I look forward to your proposal. According to our triage process, this should be tagged VNext and the title of the bug updated to reflect the more specific request.

**Comment by GravisZro** [comment]
> That would be the license server imposing the restriction, not the CDM. You can do that with or without EME.

You are working under the assumption that you have just requested the content from a network source.  If I were to download the content and view it later in another country, I would suddenly be denied because of my geographic location.

> I look forward to your proposal.

I'm not going to write up a proposal, that's not my job.  What I have done is report a flaw in EME as well as provide a possible solution.  You don't close a bug until it's resolved and you don't expect users to submit patches.

**Comment by wgenemorgan** [comment]
On Mon, Apr 11, 2016 at 3:10 PM, Gravis <notifications@github.com> wrote:

> If I were to download the content and view it later in another country, I
> would suddenly be denied because of my geographic location.


You have seen this happen?  Can you supply more details?

**Comment by GravisZro** [comment]
@wgenemorgan I've seen plenty of videos on the internet be blocked by geographic region.  There is no reason to think EME license requests wouldn't work the same way.

**Comment by mwatson2** [comment]
So, again, a request to a server (such as a license request) can be blocked by geo-filters whether or not you are using EME.

CDMs should not have access to your IP address, so if you have downloaded content for offline viewing, you should be fine wherever you are.

**Comment by wgenemorgan** [comment]
On Mon, Apr 11, 2016 at 3:33 PM, mwatson2 <notifications@github.com> wrote:

> so if you have downloaded content for offline viewing, you should be fine
> wherever you are


In my experience, this is precisely what happens on a platform using a CDM.

**Comment by GravisZro** [comment]
@mwatson2
> CDMs should not have access to your IP address, so if you have downloaded content for offline viewing, you should be fine wherever you are.

If the format requires the CDM for playback, the CDM will request the license upon playback.  It's indirect but at that point the server can deny you the license key based on your IP address.  You can say, "well that's a server issue" all you want but we know that's a lie.

@wgenemorgan 
> In my experience, this is precisely what happens on a platform using a CDM.

Which is why it's a hypothetical situation.  Always expect malicious behavior when it comes to computers.

**Comment by paulbrucecotton** [comment]
>You can say, "well that's a server issue" all you want but we know that's a lie.

@GravisZro: It is inappropriate to accuse other participants in this dialogue of including lies in their responses.  Please refrain from this kind of responses and keep you discussion to technical points.

Paul Cotton
HTML Media WG chair

**Comment by GravisZro** [comment]
@paulbrucecotton I agree the conversation should be kept technical, please remind other participants of this as well.  mwatson2 in particular has gone way off base in regards to the technical discussions.  I also request that participants be honest about the implications of the API itself.

**Comment by mwatson2** [comment]
I'm not sure where I am 'way off base', but to answer your questions, for offline viewing scenarios the CDM will need to persist the license (so that it is available offline) and so there will be no server interaction at the time of playback (obviously, there cannot be if the thing is going to work offline ).

For online viewing, services may indeed apply geographic restrictions. These are a fact of life the way content is licensed today. My point was that it is a server function to apply these restrictions, not something that is done by the DRM. This is important to recognize because there would be privacy implications if the CDM could access your location.

It is of course possible to build a service based on downloaded content but that requires online authorization for every playback. This would be a little odd, since the primary advantage of download is for viewing when you do not have connectivity or only bad connectivity. Still it is the server, not the CDM, that makes the authorization decision.

Finally, regarding interoperability, as Joe mentioned, there is nothing in the specification which prevents or discourages the implementation of cross-platform CDMs and indeed Google's solution is supported on several platforms as will be Mozilla's solution.

It's been discussed for a long time whether we could do more to promote cross-platform solutions but no one has any bright ideas that we could execute on any time soon. Hence my frustration at the implication that this is just a bug to be solved like any other.

A trusted virtual execution environment for CDMs is a nice idea, but to my knowledge there is no work available in this area (research, prototypes etc.) that we could draw on.

**Comment by GravisZro** [comment]
> regarding interoperability, as Joe mentioned, there is nothing in the specification which prevents or discourages the implementation of cross-platform CDMs

That is disingenuous because there absolutely is, the sheer number of platforms and architectures!  How many companies are going to be backporting to unsupported platforms?  OSX for PPC?  that's a large number of computers and yet it's not supported.  Then you have more obscure platforms like FreeBSD on SPARC64.  Netflix/Google/Adobe/Microsoft is not going to spend the resources to support a platform that is used by very few people when you practically have to break their arm just to get them to support Linux on x86.

> A trusted virtual execution environment for CDMs is a nice idea, but to my knowledge there is no work available in this area (research, prototypes etc.) that we could draw on.

I'm not talking about trusted execution, just a simple interpreter for a binary format that has a limited instruction set.

**Comment by GravisZro** [comment]
@mwatson2: I've given it some thought and I may have come up with a solution to the CDM portability issue.  [Standard Portable Intermediate Representation (SPIR)](https://www.khronos.org/spir) is an intermediate language for parallel compute and graphics.  It's an open standard, fully developed, has a existing free/libre compiler, is fast and immediately addresses a lot of sandboxing and computational performance issues when used with modern graphics cards.  I also believe it meets your "robustness" requirement.

**Comment by steelejoe** [comment]
@ddorwin @mwatson2  Just thought I would note that most if not all of the topics raised here are included in [bug 20944](https://www.w3.org/Bugs/Public/show_bug.cgi?id=20944) in the old bug tracker, which is still open. I am not sure how we decided to deal with those bugs, but one or more of these bugs should probably be closed as a dup.

**Comment by GravisZro** [comment]
@steelejoe it's worth noting that the other bug was last active two years ago and no progress has been made.

**Comment by paulbrucecotton** [comment]
After discussion with the Editors we have decided to open [ISSUE-192](https://github.com/w3c/encrypted-media/issues/192) based on the long outstanding [Bugzilla bug 20944](https://www.w3.org/Bugs/Public/show_bug.cgi?id=20944).

This issue will be closed as a duplicate of ISSUE-192.

**Comment by ddorwin** [comment]
Closing as duplicate per the comment above.
